File: MetodosOrdenamiento/heap/heap.py
import random
import os

def heapify(arr,n,i):
    largest = i
    left = 2*i+1
    right = 2*i+2

    if left < n and arr[left]>arr[largest]:
        largest = left

    if right < n and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        arr[i],arr[largest] = arr[largest],arr[i]
        heapify(arr, n, largest)

def heap_sort(arr):
    n = len(arr)

    for i in range(n // 2 -1, -1,-1):
        heapify(arr,n,i)

    for i in range (n-1,0,-1):
        arr[i],arr[0] = arr[0],arr[i]
        heapify(arr,i,0)

def main():
    os.system('clear')
    numeros = [random.randint(1,100) for _ in range(10)]

    print("Arreglo original:")
    print(numeros)

    numeros_ordenados = numeros.copy()

    heap_sort(numeros_ordenados)

    print("Arreglo ordenado")
    print(numeros_ordenados)

if __name__ == "__main__":
    main()

File: MetodosOrdenamiento/heap/heap.cpp
#include <iostream>
#include <vector>
#include <random>
#include <algorithm>

using namespace std;
void heapify(vector<int>& arr, int n, int i){
    int largest = i;
    int left = 2*i+1;
    int right = 2*i+2;

    if (left < n && arr[left] > arr [largest]){
        largest = left;
    }

    if (right < n && arr[right] > arr[largest]){
        largest = right;
    }

    if(largest != i){
        swap(arr[i],arr[largest]);
        heapify(arr,n,largest);
    }
} 
void heapSort(vector<int>& arr){
    int n= arr.size();

    for(int i=n/2-1;i>=0;i--){
        heapify(arr,n,i);
    }
    for(int i=n-1;i>0;i--){
        swap(arr[0], arr[i]);
        heapify(arr,i,0);
    }
}  
int main(){
    system("clear");
    vector<int> numeros;

    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dis(1,100);

    for (int i=0;i<10;i++){
        numeros.push_back(dis(gen));
    }

    cout << "Arreglo original:\n";
    for(int num : numeros){
        cout << num << " ";
    }
    cout << endl;

    vector<int> numerosOrdenados = numeros;
    heapSort(numerosOrdenados);
    
    for(int num : numerosOrdenados){
        cout << num << " ";
    }
    cout << endl;
}
